### PR TITLE
Changed track registry to be backed by forge controlled registry

### LINKED ISSUE #17: Use forge namespaced registry for track type and track kit

### PR DESCRIPTION
**Pull request by liach**
Resolves #17.
Will open a pull in Railcraft repo soon.

### CODE REVIEW

File: mods/railcraft/api/tracks/TrackRegistry.java [hunk 1]
```diff
@@ -37,142 +43,222 @@ This work (the API) is licensed under the "MIT" License,
...
     public void finalizeRegistry() {
-        if (!Loader.instance().isInState(LoaderState.PREINITIALIZATION))
-            throw new TrackRegistryException("Finalize called outside PRE-INIT");
-        List<T> list = new ArrayList<>();
-        list.addAll(registry.values());
+        if (!RailcraftConstantsAPI.MOD_ID.equals(Loader.instance().activeModContainer().getModId())) {
+            throw new TrackRegistryException("Finalize called by non-railcraft mods!");
+        }
+        List<T> list = Lists.newArrayList(iterator());
         list.sort(Comparator.comparing(T::getName));
```
Comment (CovertJaguar) [inline mods/railcraft/api/tracks/TrackRegistry.java]: list can be removed, it was used to populate idMap

**Review by CovertJaguar — CHANGES_REQUESTED** [review]

File: mods/railcraft/api/tracks/TrackRegistry.java (hunk 1 above)
Comment (liach) [reply mods/railcraft/api/tracks/TrackRegistry.java]: Didn't realize that. Will fix after updating intellij idea.

File: mods/railcraft/api/tracks/TrackRegistry.java [hunk 2]
```diff
@@ -114,9 +113,8 @@ public void finalizeRegistry() {
         if (!RailcraftConstantsAPI.MOD_ID.equals(Loader.instance().activeModContainer().getModId())) {
             throw new TrackRegistryException("Finalize called by non-railcraft mods!");
         }
-        List<T> list = Lists.newArrayList(iterator());
-        list.sort(Comparator.comparing(T::getName));
-        if (combinations.isEmpty()) {
+        pass++; // Prevent building when not all track types or track kits are registered.
+        if (pass == 2) {
```
Comment (CovertJaguar) [inline mods/railcraft/api/tracks/TrackRegistry.java]: What is this is and how does it work? Why is this called multiple times?

Comment (liach) [reply mods/railcraft/api/tracks/TrackRegistry.java]: pass 1: one registry finalizes;
pass 2: the other finalizes.
After both registries are freezed, we build the list.

Comment (CovertJaguar) [inline mods/railcraft/api/tracks/TrackRegistry.java]: So what calls it the third time?

**Review by CovertJaguar — APPROVED** [review]

**Review by CovertJaguar — CHANGES_REQUESTED** [review]
You need to bump the API version number for this, probably a major revision from the looks of the missing functions.

**Comment by liach** [comment]
Can I bump for both pulls?

**Comment by liach** [comment]
Like a bump for 2 breaking changes

**Comment by CovertJaguar** [comment]
Technically the other one, if implemented the way I say, is only an addition. But this one is breaking because you removed a method, so the Track Package info gets a major version bump.

But yeah you can consider a single bump good for both.

**Review by CovertJaguar — APPROVED** [review]

**Pull request merged**